lib/path-design.md

```diff
@@ -0,0 +1,177 @@
+- `path + ("/" + string)` returns the path to the `string` subdirectory in `path`
+  - If `string` contains no `/` characters, then `dirOf (path + ("/" + string)) == path`
+  - If `string` contains no `/` characters, then `baseNameOf (path + ("/" + string)) == string`
+- `path1 == path2` returns true only if `path1` points to the same filesystem path as `path2`
+
+Notably we do not make the assumption that we can turn paths into strings using `toString path`.
+
+## Design decisions
+
+Each subsection here contains a decision along with arguments and counter-arguments for (+) and against (-) that decision.
+
+### Leading dots for relative paths
+[leading-dots]: #leading-dots-for-relative-paths
+
+Context: Relative paths can have a leading `./` to indicate it being a relative path, this is generally not necessary for tools though
+
+Decision: Returned relative paths should always have a leading `./`
+
+<details>
+<summary>Arguments</summary>
+
+- :heavy_plus_sign: In shells, just running `foo` as a command wouldn't execute the file `foo`, whereas `./foo` would execute the file. In contrast, `foo/bar` does execute that file without the need for `./`. This can lead to confusion about when a `./` needs to be prefixed. If a `./` is always included, this becomes a non-issue. This effectively then means that paths don't overlap with command names.
+- :heavy_plus_sign: Prepending with `./` makes the relative paths always valid as Nix path expressions
+- :heavy_plus_sign: Using paths in command line arguments could give problems if not escaped properly, e.g. if a path was `--version`. This is not a problem with `./--version`. This effectively then means that paths don't overlap with GNU-style command line options
+- :heavy_minus_sign: `./` is not required to resolve relative paths, resolution always has an implicit `./` in front
+- :heavy_minus_sign: It's more pretty without the `./`, good for error messages and co.
+  - :heavy_plus_sign: But similarly, it could be confusing whether something was even a path
+    e.g. `foo` could be anything, but `./foo` is more clearly a path
+- :heavy_plus_sign: Makes it more uniform with absolute paths (those always start with `/`)
+  - :heavy_minus_sign: Not relevant though, this perhaps only simplifies the implementation a tiny bit
+- :heavy_plus_sign: `find` also outputs results with `./`
+  - :heavy_minus_sign: But only if you give it an argument of `.`. If you give it the argument `some-directory`, it won't prefix that
+- :heavy_minus_sign: `realpath --relative-to` doesn't output `./`'s
+  - :heavy_plus_sign: We don't need to return the same result though
+
+</details>
+
+### Representation of the current directory
+[curdir]: #representation-of-the-current-directory
+
+Context: The current directory can be represented with `.` or `./` or `./.`
```

Comment:
Nix does not have a concept of current directory, except in the path expression parsing stage. This is transparent to anything written in the language.
We do need a term for the first argument of the path join operation or relative path resolution operation. In a web context, this would be the "base".
We could say that the _base_ of a Nix file is the directory that contains the file, for the purpose of resolving a path expression to a path value.
Similarly we could have `lib.path.join base subpath`.

```suggestion
Context: The relative path that produces the base can be represented with `.` or `./` or `./.`
```

We might want to start the document with a terminology section. That section should be clear of any design uncertainty. We can discuss decisions made in the terminology in one of these context / decision sections.

Reply:
Applied this with "base directory"